data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java

```diff
@@ -0,0 +1,274 @@
+    private final Counter numberOfRecordsFailedCounter;
+    private final Timer lambdaLatencyMetric;
+    private final String invocationType;
+    private final Collection<EventHandle> bufferedEventHandles;
+    private final List<Event> events;
+    private final BatchOptions batchOptions;
+    private final ObjectMapper objectMapper = new ObjectMapper();
+    private final BufferFactory bufferFactory;
+    private final LambdaClient lambdaClient;
+    private final Boolean isBatchEnabled;
+    private final String batchKey;
+    Buffer currentBuffer;
+    private final AtomicLong requestPayload;
+    private final AtomicLong responsePayload;
+    private int maxEvents = 0;
+    private ByteCount maxBytes = null;
+    private Duration maxCollectionDuration = null;
+    private int maxRetries = 0;
+    private String mode = null;
+    private OutputCodec codec = null;
+
+    @DataPrepperPluginConstructor
+    public LambdaProcessor(final PluginMetrics pluginMetrics, final LambdaProcessorConfig lambdaProcessorConfig, final AwsCredentialsSupplier awsCredentialsSupplier, final ExpressionEvaluator expressionEvaluator) {
+        super(pluginMetrics);
+        this.expressionEvaluator = expressionEvaluator;
+        this.numberOfRecordsSuccessCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_FLUSHED_TO_LAMBDA_SUCCESS);
+        this.numberOfRecordsFailedCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_FLUSHED_TO_LAMBDA_FAILED);
+        this.lambdaLatencyMetric = pluginMetrics.timer(LAMBDA_LATENCY_METRIC);
+        this.requestPayload = pluginMetrics.gauge(REQUEST_PAYLOAD_SIZE, new AtomicLong());
+        this.responsePayload = pluginMetrics.gauge(RESPONSE_PAYLOAD_SIZE, new AtomicLong());
+        functionName = lambdaProcessorConfig.getFunctionName();
+        whenCondition = lambdaProcessorConfig.getWhenCondition();
+        maxRetries = lambdaProcessorConfig.getMaxConnectionRetries();
+        batchOptions = lambdaProcessorConfig.getBatchOptions();
+        if (batchOptions != null) {
+            maxEvents = batchOptions.getThresholdOptions().getEventCount();
+            maxBytes = batchOptions.getThresholdOptions().getMaximumSize();
+            maxCollectionDuration = batchOptions.getThresholdOptions().getEventCollectTimeOut();
+            batchKey = batchOptions.getBatchKey();
+            isBatchEnabled = true;
+            LOG.info("maxEvents:" + maxEvents + " maxbytes:" + maxBytes + " maxDuration:" + maxCollectionDuration);
+        } else {
+            batchKey = null;
+            isBatchEnabled = false;
+        }
+        mode = lambdaProcessorConfig.getMode();
+        // TODO - Support for Async mode to be added.
+        if (mode != null && mode.equalsIgnoreCase(LambdaProcessorConfig.SYNCHRONOUS_MODE)) {
+            invocationType = SYNC_INVOCATION_TYPE;
+        } else {
+            throw new RuntimeException("Unsupported mode " + mode);
+        }
+
+        codec = new LambdaJsonCodec(batchKey);
```

Comment:
This approach is very restrictive. It assumes that the body fits the format we ask. We should follow the same pattern used elsewhere in Data Prepper by allowing for a pluggable output codec.

Reply:
LambdaJsonCodec is an implementation of OutputCodec - [link](https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/common/codec/LambdaJsonCodec.java#L20) . I needed this specifically for batch processing when more than one event needs to be mapped to json. JsonOutputCodec is only event at a time. Maybe i can change the name to something more generic like BulkJsonOutputCodec? It will be used for all dial out processors.

Reply:
All `OutputCodec`s are made for batches.

Customers can use either `JsonOutputCodec` or `NdjsonOutputCodec`. 

The `json` codec already supports a configurable key name. This would replace the `batch_key` which you don't need.

```
codec:
  json:
    key_name: myKey
```

Reply:
So the customer can have two options:

1. JSON

```
codec:
  json:
    key_name: myKey
```

Yields:

```
{
  "myKey" : [
     { ...event1... },
     { ...event2... },
     { ...event3... }
  ]
}
```

2. The customer can use `ndjson`

```
codec:
  ndjson:
```

Yields:

```
{ ...event1... }
{ ...event2... }
{ ...event3... }
```

Reply:
Understood, there is also a difference when it comes to handling single event without batch. In this case, i still want to convert dataprepper event to json but i dont want to have a key, i want to pass on the user's data as it is to lambda as payload; but current output codec forces me to have a key. To address that, i either need to add new behaviour to json and ndjson writeEvent method, to convert event directly to json OR write a new codec(which is what i did).

[LambdaJsonCodec](https://github.com/srikanthjg/data-prepper/blob/8f33225a32bc4ba226f6f83eb88d3ffff5212ad6/data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/common/codec/LambdaJsonCodec.java#L62) and [JsonOutputCodec](https://github.com/srikanthjg/data-prepper/blob/8f33225a32bc4ba226f6f83eb88d3ffff5212ad6/data-prepper-plugins/parse-json-processor/src/main/java/org/opensearch/dataprepper/plugins/codec/json/JsonOutputCodec.java#L70) difference in the way we handle writeEvent. 

If we want to go with the pluggable codec approach, i think we need to modify the behaviour of the existing outputCodec, is that what we want?

Reply:
regarding the configuration, i already have "invocation_type" as a configuration, this allows to set per event invocation or batch invocation(RequestResponse or Event).

If we are implementing this internally, i cannot use parse-json-processor as a plugin but will have to take a dependency on it. Is it ok for one processor to take a dependency on the other?

I wanted to avoid this, hence went with implementing a custom codec. But i think this codec can also be used by other dial-out processors eventually, i can make it generic.